feat(providers): Add Google Drive support.

xkxx opened the pull request:
Hi @jkowalski,

I'm requesting early review for this feature before I finish up with tests and documentation. This PR implements https://github.com/kopia/kopia/issues/90 and passes `validate-provider`.

The implementation strategy is the one we discussed in https://github.com/kopia/kopia/issues/90#issuecomment-1025507405:
* FileIdCache: A BlobId -> FileId map is put in place to speed up file look ups.
* GetBlob / GetBlobMetadata: Use exact name match with aggressive retrying (100ms x 3 times). 
* ListBlobs: Use name prefix search with a cap of 26 characters (Drive limitation). Additionally, we search through FileIdCache for any local matches in case the Drive-side index is not stable yet.

I would like your input on:
* The general implementation strategy - is it sensible? Don't worry about go readability right now as I can clean it up after the big picture is set.
* Right now, FileIdCache is an ephemeral and uncapped cache. The fact is that this cache is effectively append-only since the id mapping is stable. Would it be better to turn it into a LRU cache or persist it on disk?

List of changes in PR:
* `repo/blob/gdrive`: Implement blob storage in GDrive.
* `cli/`: Implement `gdrive` storage subcommands.
* `internal/retry/retry.go`: Fix a bug affecting gdrive where `Errorf` is used instead of `Wrapf`, obscuring the error chain.

Thanks,

Kexiang

jkowalski commented on repo/blob/gdrive/gdrive_storage.go:
```diff
@@ -0,0 +1,460 @@
+	client      *drive.Service
+	folderId    string
+	// Caches blob id -> drive file id associations.
+	fileIdCache map[blob.ID]string
```
can you define a separate struct type for the fileIDCache so we can test unit-test all the cache operations?

xkxx replied:
That's a good idea. I also used the chance to add synchronization so that all the concurrent tests pass now.

jkowalski commented on repo/blob/gdrive/gdrive_storage.go:
```diff
@@ -0,0 +1,460 @@
+type gdriveStorage struct {
+	Options
+
+	client      *drive.Service
```
I don't know if drive API has something that prevents mocking, but I'm hoping `client` be an interface includes just the methods of `*drive.FilesService` we actually use?

This would allow us to write fake implementation of this interface and unit-test most of this provider without calling actual Google drive, because unlike others it has some pretty non-trivial logic?

I realize this is can be much harder than the current approach (so it perhaps can be done in separate PR, once this early version is merged), but I think it's important to prevent regressions here.

xkxx replied:
Looks like FilesService doesn't mind being stored on its own. Done

jkowalski commented on repo/blob/gdrive/gdrive_storage.go:
```diff
@@ -0,0 +1,460 @@
+
+	fileId, err := gdrive.GetFileId(ctx, b)
+	if err != nil {
+		return err
```
in general prefer `errors.Wrap()` over `return err` - this helps with tracking down sources of errors.

xkxx replied:
Thanks for the tip. I went through and added Wrapf to all the places where we interface with drive api.

jkowalski commented on repo/blob/gdrive/gdrive_storage.go:
```diff
@@ -0,0 +1,460 @@
+	consumer := func(files *drive.FileList) error {
+		for _, file := range files.Files {
+			blobId := toBlobId(file.Name)
+			gdrive.fileIdCache[blobId] = file.Id
```
cache must be thread/goroutine safe, so you'd need a mutex inside cache implementation - simple map access will generally panic on concurrent access attempts.

xkxx replied:
Done.

jkowalski commented on repo/blob/gdrive/gdrive_storage.go:
```diff
@@ -0,0 +1,460 @@
+
+func (gdrive *gdriveStorage) ListBlobs(ctx context.Context, prefix blob.ID, callback func(blob.Metadata) error) error {
+	// Tracks blob matches in cache but not returned by API.
+	unvisitedIds := make(map[blob.ID]bool)
```
n/b: this is probably fine for most use cases, but maps like this can get really big for large repositories potentially using lots of RAM.

xkxx replied:
I decided to split this up to two caches:
- Blobs: This cache tracks blob -> file mapping and is unbounded.
- ChangeLog: This cache tracks recent blob creations / deletions and is limited to 2^8 entries. Only this cache is iterated through for ListBlobs.

jkowalski commented on repo/blob/gdrive/gdrive_storage.go:
```diff
@@ -0,0 +1,460 @@
+
+	query := fmt.Sprintf("'%s' in parents and mimeType = '%s'", gdrive.folderId, gdriveMimeType)
+	if prefix != "" {
+		query = fmt.Sprintf("'%s' in parents and name contains '%s' and mimeType = '%s'", gdrive.folderId, capPrefix(prefix), gdriveMimeType)
```
should it be `contains`? is there 'startsWith' or something like that?

xkxx replied:
So according to the drive api `contains` acts like `startswith` for file names. I added a comment here to clarify this.

jkowalski commented:
Overall strategy LGTM, i did not comment on linter errors since those are already included.

I would also recommend cloing `gcs_storage_test.go` and running tests locally to make sure the required semantics is met by this provider. When you do that, expect all credentials to be passed via environment variables (see `provider-tests.yaml`) - this way we will be able to automate test runs against this provider on CI/CD and you can easily pass those when developing locally.

xkxx commented:
Thanks for the review!

I added a new commit with these changes:
- Added a synchronized FileIdCache. As you pointed out, this is necessary for concurrent repo access. This cache has two parts:
  - Blobs: This tracks individual blob -> file mappings. Individual rows have their own mutex locks. Since all lookups can involve writes, this is a mutex not rwmutex.
  - ChangeLog: This tracks recent changes for ListBlobs. One rwmutex governs this.
- Added tests and make sure all of these pass locally.
- Fix lint and style issues.

xkxx commented on repo/blob/gdrive/gdrive_storage.go:
```diff
@@ -0,0 +1,556 @@
+	req := gdrive.client.Get(fileID)
+	req.Header().Set("Range", toRange(offset, length))
+	res, err := req.Context(ctx).Download()
+	
```
There's an issue with the linter setup:
- gofumpt wants to remove it
- wsl wants to add it back

jkowalski replied:
i think you need to add empty line above 66

xkxx replied:
That fixed it. I apparently misread the error output. Thanks!

jkowalski commented on repo/blob/gdrive/file_id_cache.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// CacheEntry is a blob cache entry.
+type CacheEntry struct {
```
nit, can you make this non-exported (lowercase)?

xkxx replied:
Done. Thanks for teaching me on go :)

jkowalski commented on repo/blob/gdrive/file_id_cache.go:
```diff
@@ -0,0 +1,130 @@
+)
+
+// FileIDCache is a cache for managing the association of blobID -> fileID.
+type FileIDCache struct {
```
nit, can you make this non-exported (lowercase)?

xkxx replied:
Done

jkowalski commented on repo/blob/gdrive/file_id_cache.go:
```diff
@@ -0,0 +1,130 @@
+}
+
+// ChangeEntry is a blob change entry.
+type ChangeEntry struct {
```
nit, can you make this non-exported (lowercase)?

xkxx replied:
Done

jkowalski commented:
I tested the PR on my local machine and it appears to be working against test gdrive and credentials I generated, which is great. I added couple CI-related changes to your branch.

I'm going to change this PR to merge into `test/providers` and we will merge to `master` when the tests pass.

One more thing we should do before merging - ensure the test can run in parallel, as of now we use shared folder, can the test be modified to create a subfolder with a random name and run with it, so that two tests don't collide with each other?

xkxx commented:
Sounds great! I added the changes on test isolation. Once this is merged, I can work on updating the documentation.

jkowalski commented:
I ran this locally and appears to all work, I'm going to merge into `test/providers` so we can get some official runs against it.

I have some concerns about retry loop in `tryGetFileByBlobID` - is this needed at all? I removed it and all the tests are passing - this retry loop appears to be slowing down tests, so if it's not needed I'd love to remove it (separate PR is fine, we can also add docs).

jkowalski commented:
Looks like the test passed on GitHub: https://github.com/kopia/kopia/runs/5227924251?check_suite_focus=true

I'm going to merge this into `master` now and we can iterate over there.

xkxx commented:
Woohoo! Thanks for guiding me through this PR!

RE tryGetFileByBlobID: It depends on how important it is to have concurrent access by multiple clients. Due to Drive indexing being eventually consistent, looking up a blob right after it's created may not work. For single client usage, this issue is resolved with `fileIdCache`. For multi clients, the retry loop mitigates the issue by essentially adding more latency to the lookup.

In the majority of cases, I don't think concurrent mutations of the same blobs (save for the repository file) is a concern. If that's the case, we can remove the retry loop and get some performance back.
